publish-mqtt/src/main.rs

```diff
@@ -303,44 +407,61 @@ async fn disconnect_first_stale_sensor(
             sensor.name,
             now - sensor.last_update_timestamp
         );
-        homie.remove_node(&sensor.node_id()).await?;
+        homie
+            .remove_node(&sensor.node_id())
+            .await
+            .with_context(|| std::line!().to_string())?;
         sensors_to_connect.push_back(sensor);
     }
     Ok(())
 }
 
 async fn service_bluetooth_event_queue(
-    bt_session: &BluetoothSession,
-    homie: &mut HomieDevice,
-    sensors_connected: &mut Vec<Sensor>,
-    sensors_to_connect: &mut VecDeque<Sensor>,
-) -> Result<(), Box<dyn Error>> {
+    state: Arc<Mutex<SensorState>>,
+    dbus_conn: Arc<SyncConnection>,
+) -> Result<(), anyhow::Error> {
+    println!("Subscribing to events");
+    let mut rule = dbus::message::MatchRule::new();
+    rule.msg_type = Some(dbus::message::MessageType::Signal);
+    rule.sender = Some(dbus::strings::BusName::new("org.bluez").map_err(|s| anyhow::anyhow!(s))?);
+
+    let (msg_match, mut events) = dbus_conn
+        .add_match(rule)
+        .await
+        .with_context(|| std::line!().to_string())?
+        .msg_stream();
+    println!("Processing events");
     // Process events until there are none available for the timeout.
-    for event in bt_session
-        .incoming(INCOMING_TIMEOUT_MS)
-        .filter_map(BluetoothEvent::from)
-    {
-        handle_bluetooth_event(homie, sensors_connected, sensors_to_connect, event).await?;
+    while let Some(raw_event) = events.next().await {
+        if let Some(event) = BluetoothEvent::from(raw_event) {
+            handle_bluetooth_event(state.clone(), event)
+                .await
+                .with_context(|| std::line!().to_string())?
+        }
     }
+    // TODO: move this into a defer or scope guard or something.
+    dbus_conn.remove_match(msg_match.token()).await?;
     Ok(())
 }
 
 async fn handle_bluetooth_event(
-    homie: &mut HomieDevice,
-    sensors_connected: &mut Vec<Sensor>,
-    sensors_to_connect: &mut VecDeque<Sensor>,
+    state: Arc<Mutex<SensorState>>,
     event: BluetoothEvent,
-) -> Result<(), Box<dyn Error>> {
+) -> Result<(), anyhow::Error> {
+    let state = &mut *state.lock().await;
+    let homie = &mut state.homie;
+    let sensors_connected = &mut state.sensors_connected;
+    let sensors_to_connect = &mut state.sensors_to_connect;
     match event {
         BluetoothEvent::Value { object_path, value } => {
             // TODO: Make this less hacky.
-            let device_path = match object_path.strip_suffix(SERVICE_CHARACTERISTIC_PATH) {
+            let object_path = match object_path.strip_suffix(SERVICE_CHARACTERISTIC_PATH) {
```

Comment:
- [ ] I prefer the name `Sensor.object_path` to mean "the object path for the root of the sensor", but in this specific function let's switch it back to device_path.